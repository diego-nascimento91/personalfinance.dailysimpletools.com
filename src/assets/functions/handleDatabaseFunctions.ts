import FirebaseFirestoreService from 'assets/functions/FirebaseFirestoreService';
import { IAccount, ICategory, IOrderConfig, IQuery, ITransaction } from 'assets/interfaces/interfaces';
import { SetterOrUpdater } from 'recoil';

export const handleFetchCategories = async (setCategories: SetterOrUpdater<ICategory[]>, userId: string) => {
  try {
    // default categories
    const collectionPathDefault = 'categories';
    const orderConfigDefault: IOrderConfig[] = [{ fieldName: 'ordering', orderDirection: 'asc' }];
    const categoriesDefault = await FirebaseFirestoreService.readAllDocsFromCollection(collectionPathDefault, orderConfigDefault);

    // user categories
    const collectionPathUser = `users/${userId}/categories`;
    const orderConfigUser: IOrderConfig[] = [{ fieldName: 'name', orderDirection: 'asc' }];
    const categoriesUser = await FirebaseFirestoreService.readAllDocsFromCollection(collectionPathUser, orderConfigUser);

    const categoriesDB = [...categoriesUser, ...categoriesDefault];
    setCategories(categoriesDB as ICategory[]);
  } catch (error) {
    if (error instanceof Error) {
      alert(error.message);
      throw error;
    }
  }
};

export const handleFetchOnlyUserCategories = async (setCategories: SetterOrUpdater<ICategory[]>, userId: string) => {
  try {
    // user categories
    const collectionPathUser = `users/${userId}/categories`;
    const orderConfigUser: IOrderConfig[] = [
      { fieldName: 'name', orderDirection: 'asc' },
      { fieldName: 'type', orderDirection: 'desc' }];
    const categoriesUser = await FirebaseFirestoreService.readAllDocsFromCollection(collectionPathUser, orderConfigUser);

    setCategories(categoriesUser as ICategory[]);
  } catch (error) {
    if (error instanceof Error) {
      alert(error.message);
      throw error;
    }
  }
};

export const handleFetchAccounts = async (setAccounts: SetterOrUpdater<IAccount[]>, userId: string) => {
  try {
    const collectionPath = `users/${userId}/accounts`;
    const orderConfig: IOrderConfig[] = [{ fieldName: 'name', orderDirection: 'asc' }];
    const accountsDB = await FirebaseFirestoreService.readAllDocsFromCollection(collectionPath, orderConfig);
    setAccounts(accountsDB as IAccount[]);
  } catch (error) {
    if (error instanceof Error) {
      alert(error.message);
      throw error;
    }
  }
};

export const handleFetchRecentTransactions = async (userId: string, setRecentTransactions: SetterOrUpdater<ITransaction[]>) => {
  const collectionPath = `users/${userId}/transactions`;
  const orderConfig: IOrderConfig[] = [
    {
      fieldName: 'date',
      orderDirection: 'desc'
    },
    {
      fieldName: 'publishDate',
      orderDirection: 'desc'
    },
  ];
  const queries: IQuery[] = [];
  const limitDocs = 4;

  try {
    const recentTransactions = await FirebaseFirestoreService.readAllDocsFromCollection(collectionPath, orderConfig, queries, limitDocs);
    setRecentTransactions(recentTransactions as ITransaction[]);
  } catch (error) {
    if (error instanceof Error) {
      alert(error.message);
      throw error;
    }
  }
};

export const handleFetchTransactionsMonth = async (userId: string, setTransactionsMonth: SetterOrUpdater<ITransaction[]>, month: Date) => {
  const collectionPath = `users/${userId}/transactions`;
  const orderConfig: IOrderConfig[] = [
    {
      fieldName: 'date',
      orderDirection: 'desc'
    },
    {
      fieldName: 'publishDate',
      orderDirection: 'desc'
    },
  ];
  const { firstDay, lastDay } = getFormatDate(month);
  const queries = getQueries(firstDay, lastDay);

  try {
    const transactionsMonth = await FirebaseFirestoreService.readAllDocsFromCollection(collectionPath, orderConfig, queries);
    setTransactionsMonth(transactionsMonth as ITransaction[]);
  } catch (error) {
    if (error instanceof Error) {
      alert(error.message);
      throw error;
    }
  }
};

export const handleCreateDocFunction = async (collectionName: string, userId: string, document: ITransaction | ICategory | IAccount) => {
  try {
    const collectionPath = `users/${userId}/${collectionName}`;
    await FirebaseFirestoreService.createDocument(collectionPath, document);
    alert('Document added successfully!');
  } catch (error) {
    if (error instanceof Error) {
      alert(error.message);
      throw error;
    }
  }
};

export const handleCreateDocsTransferFunction = async (collectionName: string, userId: string, documentFrom: ITransaction, documentTo: ITransaction) => {
  try {
    const collectionPath = `users/${userId}/${collectionName}`;
    const documentFromRef = await FirebaseFirestoreService.createDocument(collectionPath, documentFrom);
    if (documentFromRef) {
      const documentToRef = await FirebaseFirestoreService.createDocument(collectionPath, { ...documentTo, transferedTransactionID: documentFromRef.id });
      if (documentToRef)
        await FirebaseFirestoreService.createDocument(collectionPath, { ...documentFrom, transferedTransactionID: documentToRef.id }, documentFromRef.id);
    }
    alert('Transfer transaction added successfully!');
  } catch (error) {
    if (error instanceof Error) {
      alert(error.message);
      throw error;
    }
  }
};

export const handleUpdateDocFunction = async (collectionName: string, userId: string, document: ITransaction | ICategory | IAccount, dontShowAlert?: boolean) => {

  try {
    const collectionPath = `users/${userId}/${collectionName}`;
    if (document.id) {
      const docId = document.id;
      delete document.id;
      await FirebaseFirestoreService.updateDocument(collectionPath, document, docId);
      if (!dontShowAlert) alert('Document updated successfully!');
    } else {
      throw new Error('Missing document ID');
    }
  } catch (error) {
    if (error instanceof Error) {
      alert(error.message);
      throw error;
    }
  }
};

export const handleDeleteDocFunction = async (collectionName: string, userId: string, document: ITransaction | ICategory | IAccount) => {
  try {
    const deleteConfirmation = window.confirm(`Are you sure you want to delete this document from your ${collectionName}? Ok for Yes. Cancel for No.`);
    if (deleteConfirmation) {
      const collectionPath = `users/${userId}/${collectionName}`;
      if (document.id) {
        await FirebaseFirestoreService.deleteDocument(collectionPath, document.id);
        alert('Document deleted successfully!');
      } else {
        throw new Error('Missing document ID');
      }
    } else {
      throw new Error('Document deletion cancelled');
    }
  } catch (error) {
    if (error instanceof Error) {
      alert(error.message);
      throw error;
    }
  }
};

export const handleDeleteDocsTransferFunction = async (collectionName: string, userId: string, docFromID: string, docToID: string) => {
  const deleteConfirmation = window.confirm('Are you sure you want to delete this transfer transaction? Ok for Yes. Cancel for No. This action will delete the two transactions related.');
  if (deleteConfirmation) {
    try {
      const collectionPath = `users/${userId}/${collectionName}`;
      await FirebaseFirestoreService.deleteDocument(collectionPath, docFromID);
      await FirebaseFirestoreService.deleteDocument(collectionPath, docToID);
      alert('Transactions deleted successfully!');

    } catch (error) {
      if (error instanceof Error) {
        alert(error.message);
        throw error;
      }
    }
  }
};

const getFormatDate = (date: Date) => {
  const month = date.getMonth();
  const year = date.getFullYear();
  const firstDay = new Date(year, month);
  const lastDay = new Date(year, month + 1, 0);

  return { firstDay, lastDay };
};

const getQueries = (firstDay: Date, lastDay: Date) => {
  const queries = [];
  queries.push({
    field: 'date',
    condition: '>=',
    value: firstDay
  });
  queries.push({
    field: 'date',
    condition: '<=',
    value: lastDay
  });
  return queries;
};