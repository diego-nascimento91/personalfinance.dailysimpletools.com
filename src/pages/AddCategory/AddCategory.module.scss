@import 'assets/styles/variables.scss';

.addCategoryPage__container {
  section {
    max-width: 50rem;
    margin-left: auto;
    margin-right: auto;
    width: 50rem;
  }
}