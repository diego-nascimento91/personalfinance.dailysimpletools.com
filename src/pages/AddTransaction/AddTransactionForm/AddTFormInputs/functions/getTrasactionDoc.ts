import { ITransaction, ITransactionType } from 'assets/interfaces/interfaces';

export const getTransactionDoc = (name: string, transactionType: ITransactionType, amount: number, transactionDate: string, account: string, notes: string, category?: string) => {

  const transaction: ITransaction = {
    description: name,
    type: transactionType,
    amount: amount,
    category: category ? JSON.parse(category).value : 'withdraw',
    date: new Date(transactionDate.replace(/-/g, '/')), //replace '-' per '/' makes the date to be created in the user timezone, instead of UTC
    account: JSON.parse(account),
    note: notes,
    publishDate: new Date(),
  };

  return transaction;
};