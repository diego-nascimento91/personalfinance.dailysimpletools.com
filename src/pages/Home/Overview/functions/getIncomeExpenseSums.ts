import { ITransaction } from 'assets/interfaces/interfaces';

export const getIncomeExpenseSums = (transactions: ITransaction[]) => {
  let sumIncome;
  let sumExpenses;

  if (transactions && transactions.length > 0 && transactions[0].id !== '') {
    //income
    const totalIncome = transactions.filter(item => (item.type === 'income')).map(item => (item.amount)).reduce(
      (previousValue, currentValue) => previousValue + currentValue, 0);
    sumIncome = totalIncome;

    //expense
    const totalExpense = transactions.filter(item => (item.type === 'expense' && item.category !== 'Credit Card Bill')).map(item => (item.amount)).reduce(
      (previousValue, currentValue) => previousValue + currentValue, 0);
    sumExpenses = totalExpense;

  } else {
    sumIncome = 0;
    sumExpenses = 0;
  }

  return [sumIncome, sumExpenses];
};