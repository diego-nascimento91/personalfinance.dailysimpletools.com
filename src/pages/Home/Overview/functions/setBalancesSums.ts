import { IAccount, ITransaction } from 'assets/interfaces/interfaces';
import { SetterOrUpdater } from 'recoil';

export const setBalancesSums = (accounts: IAccount[], transactions: ITransaction[],setSumCurrentBalance: SetterOrUpdater<number>, setSumProjectedBalance: SetterOrUpdater<number>) => {

  if (accounts && accounts.length > 0 && accounts[0].id !== '') {
    //balance account
    const totalBalanceAccount = accounts.filter(item => (item.type === 'balance-account')).map(item => (item.balance)).reduce(
      (previousValue, currentValue) => previousValue + currentValue, 0);
    setSumCurrentBalance(totalBalanceAccount);

    //credit account
    const totalCreditAccount = transactions.filter(item => (item.account.type === 'credit-account')).map(item => (item.amount)).reduce(
      (previousValue, currentValue) => previousValue + currentValue, 0);
    setSumProjectedBalance(totalBalanceAccount - totalCreditAccount);

  } else {
    setSumCurrentBalance(0);
    setSumProjectedBalance(0);
  }
};