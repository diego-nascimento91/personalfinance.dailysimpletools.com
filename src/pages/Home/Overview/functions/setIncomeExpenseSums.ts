import { ITransaction } from 'assets/interfaces/interfaces';
import { SetterOrUpdater } from 'recoil';

export const setIncomeExpenseSums = (transactions: ITransaction[], setSumIncome: SetterOrUpdater<number>, setSumExpense: SetterOrUpdater<number>) => {
  if (transactions && transactions.length > 0 && transactions[0].id !== '') {
    //income
    const totalIncome = transactions.filter(item => (item.type === 'income')).map(item => (item.amount)).reduce(
      (previousValue, currentValue) => previousValue + currentValue, 0);
    setSumIncome(totalIncome);

    //expense
    const totalExpense = transactions.filter(item => (item.type === 'expense' && item.category !== 'Credit Card Bill')).map(item => (item.amount)).reduce(
      (previousValue, currentValue) => previousValue + currentValue, 0);
    setSumExpense(totalExpense);

  } else {
    setSumIncome(0);
    setSumExpense(0);
  }
};