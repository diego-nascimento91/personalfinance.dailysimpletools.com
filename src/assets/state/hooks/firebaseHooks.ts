import { authenticated } from 'assets/functions/FirebaseConfig';
import { useAuthState } from 'react-firebase-hooks/auth';
import { useRecoilState } from 'recoil';
import { useCategoriesAtom, useChosenMonthAtom, useTransactionsAllAtom, useTransactionsMonthAtom } from '../firebaseAtoms';

export const useUser = () => {
  return useAuthState(authenticated);
};

export const useCategories = () => {
  return useRecoilState(useCategoriesAtom);
};

export const useChosenMonth = () => {
  return useRecoilState(useChosenMonthAtom);
};

export const useTransactionsAll = () => {
  return useRecoilState(useTransactionsAllAtom);
};

export const useTransactionsMonth = () => {
  return useRecoilState(useTransactionsMonthAtom);
};