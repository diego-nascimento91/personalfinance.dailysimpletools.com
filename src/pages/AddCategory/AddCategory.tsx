import { useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { useUser } from 'assets/state/hooks/firebaseHooks';
import { handleFetchOnlyUserCategories } from 'assets/functions/handleDatabaseFunctions';
import { useSelectedCategory, useUserCategories } from 'assets/state/hooks/addCategoryHooks';
import styles from './AddCategory.module.scss';
import stylesPadding from 'assets/styles/padding.module.scss';
import stylesPages from 'assets/styles/pages.module.scss';
import AddCategoryForm from './AddCategoryForm/AddCategoryForm';
import UserCategories from './UserCategories/UserCategories';

const AddCategory = () => {
  const nav = useNavigate();
  const [user, loading] = useUser();
  const [, setUserCategories] = useUserCategories();
  const [, setSelectedCategory] = useSelectedCategory();

  useEffect(() => {
    if (loading) return;
    if (!user) nav('/');
    if(user) handleUpdateDBs();
  }, [user]);

  useEffect(() => {
    if(user) setSelectedCategory(null);
  },[]);

  const handleUpdateDBs = () => {
    if(user) {
      handleFetchOnlyUserCategories(setUserCategories, user.uid);
    }
  };

  return (
    <div className={`${stylesPages.pages} ${stylesPadding.padding} ${styles.addCategoryPage__container}`}>
      <div>
        <AddCategoryForm />
        <UserCategories />
      </div>
    </div>
  );
};

export default AddCategory;