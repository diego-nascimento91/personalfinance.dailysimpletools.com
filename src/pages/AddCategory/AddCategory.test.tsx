/* eslint-disable react/display-name */
import { BrowserRouter } from 'react-router-dom';
import { RecoilRoot } from 'recoil';
import { render, screen } from '@testing-library/react';
import { useUser } from 'assets/state/hooks/firebaseHooks';
import { handleFetchOnlyUserCategories } from 'assets/functions/handleDatabaseFunctions';
import AddCategory from './AddCategory';

// useUser mock
jest.mock('assets/state/hooks/firebaseHooks', () => ({
  ...jest.requireActual('assets/state/hooks/firebaseHooks'),
  useUser: jest.fn()
}));

// useSelectedCategory and useUserCategories mock
const mockedSetUserCategories = jest.fn();
const mockedSetSelectedCategory = jest.fn();
jest.mock('assets/state/hooks/addCategoryHooks', () => ({
  useSelectedCategory: () => ([[], mockedSetSelectedCategory]),
  useUserCategories: () => ([[], mockedSetUserCategories])
}));

// useNavigate mock
const mockedNavegacao = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedNavegacao
}));

// handleFetchOlyUserCategories mock
jest.mock('assets/functions/handleDatabaseFunctions');

// React Components mocks
jest.mock('./AddCategoryForm/AddCategoryForm', () => () => (<div>AddCategoryForm</div>));
jest.mock('./UserCategories/UserCategories', () => () => (<div>UserCategories</div>));

describe('<AddCategory /> Page', () => {
  describe('User not LoggedIn', () => {
    beforeEach(() => {
      (useUser as jest.Mock).mockReturnValue([false, false]);
    });

    it('should return to login page', () => {
      render(
        <RecoilRoot>
          <BrowserRouter>
            <AddCategory />
          </BrowserRouter>
        </RecoilRoot>
      );

      expect(mockedNavegacao).toBeCalled();
      expect(mockedNavegacao).toBeCalledWith('/');
    });
  });

  describe('User LoggedIn', () => {
    const user = { uid: 'uid' };
    beforeEach(() => {
      (useUser as jest.Mock).mockReturnValue([user, false]);
    });

    it('should stay in this page', () => {
      render(
        <RecoilRoot>
          <BrowserRouter>
            <AddCategory />
          </BrowserRouter>
        </RecoilRoot>
      );

      expect(mockedNavegacao).not.toBeCalled();
    });

    it('should call setSelectedCategory()', () => {
      render(
        <RecoilRoot>
          <BrowserRouter>
            <AddCategory />
          </BrowserRouter>
        </RecoilRoot>
      );

      expect(mockedSetSelectedCategory).toBeCalled();
      expect(mockedSetSelectedCategory).toBeCalledWith(null);
    });

    it('should call handleFetchOnlyUserCategories()', () => {
      render(
        <RecoilRoot>
          <BrowserRouter>
            <AddCategory />
          </BrowserRouter>
        </RecoilRoot>
      );

      expect(handleFetchOnlyUserCategories).toBeCalled();
      expect(handleFetchOnlyUserCategories).toBeCalledWith(mockedSetUserCategories, user.uid);
    });

    // This test will guarantee that if the user state changes to false, then the page will be redirected to '/'
    // it('should return to login page if user signs out', async () => {
    //   (useUser as jest.Mock).mockReturnValueOnce([true, false]).mockReturnValueOnce([false,false]);
    //   render(
    //     <RecoilRoot>
    //       <BrowserRouter>
    //         <AddCategory />
    //       </BrowserRouter>
    //     </RecoilRoot>
    //   );

    //   expect(mockedNavegacao).toBeCalled();
    // });

    it('should render <AddCategoryForm />', () => {
      render(
        <RecoilRoot>
          <BrowserRouter>
            <AddCategory />
          </BrowserRouter>
        </RecoilRoot>
      );

      const AddCategoryForm_Component = screen.getByText('AddCategoryForm');

      expect(AddCategoryForm_Component).toBeInTheDocument();
    });
    
    it('should render <UserCategories />', () => {
      render(
        <RecoilRoot>
          <BrowserRouter>
            <AddCategory />
          </BrowserRouter>
        </RecoilRoot>
      );

      const AddCategoryForm_Component = screen.getByText('UserCategories');

      expect(AddCategoryForm_Component).toBeInTheDocument();
    });

    it('should render <AddCategory/> correctly', () => {
      const { container } = render(
        <RecoilRoot>
          <AddCategory />
        </RecoilRoot>
      );
  
      expect(container).toMatchSnapshot();
    });
  });
});