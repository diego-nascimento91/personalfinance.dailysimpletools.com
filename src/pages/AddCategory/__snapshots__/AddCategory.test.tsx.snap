// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddCategory /> Page User LoggedIn should render <AddCategory/> correctly 1`] = `
<div>
  <div
    class="pages padding addCategoryPage__container"
  >
    <div>
      <div>
        AddCategoryForm
      </div>
      <div>
        UserCategories
      </div>
    </div>
  </div>
</div>
`;
